
当服务的提供者宕机时，消费者无法正确访问。
通常情况下，消费者需要==等待 连接超时后==，才会给出失败提示。
在spring cloud中，有一个断路器的组件，它可以实现快速的失败相应。
如果断路器在一段时间内，检测到许多类似的错误（比如超时），它就会让消费者快速失败的模块，而不在请求所依赖的服务。这样可以防止服务雪崩。
断路器也可以感应到我们的程序是否已经修正错误，如果程序已经修正，则会让消费者调用正确的服务。


以下演示通过consumer，也是ribbon项目
# 实验1 Ribbon项目使用断路器
首先加入断路器配置的依赖
![[Pasted image 20241226231905.png]]

启动类加注解
![[Pasted image 20241226232054.png]]

此时我们需要在controller中调远程服务的方法上指定一个请求失败的方法，
失败（降级）方法的必须与该方法拥有同样的参数
![[Pasted image 20241226232411.png]]


# 实验2 Feign 中使用断路器

Feign中自带了断路器，直接开启就可以了，不用引入依赖（牛逼，自带Ribbon，还自带断路器Hystrix）
![[Pasted image 20241226232720.png]]

创建service实现类，这个实现类的方法就是接口调用失败时走的代码
![[Pasted image 20241226232858.png]]

在controller中指定 出现错误 处理请求的类，就上面的实现类
![[Pasted image 20241226233024.png]]

结束



==断路器还有一个仪表盘可以直接查看服务对应的信息==
打开服务的消费者Feign项目，feign的项目者和Ribbon的消费者可以进行同样的配置

引入依赖
![[Pasted image 20241226233348.png]]

启动类添加注解
![[Pasted image 20241226233427.png]]

配置文件中开启仪表盘
![[Pasted image 20241226233446.png]]

访问服务+端口+hystrix 可以看到仪表盘界面
![[Pasted image 20241226233517.png]]

![[Pasted image 20241226233626.png]]